File: ports/dotnet/src/RussianDocs.DocumentProcessing/Postprocess/Detector.cs
using RussianDocs.DocumentProcessing.Tensors;

namespace RussianDocs.DocumentProcessing.Postprocess;

/// <summary>One detection. Mutable, because the coordinates are rewritten in place when mapped back.</summary>
public sealed class Box
{
    public double X1, Y1, X2, Y2;
    public double Conf;
    public int Cls;
    public string Label = "";

    /// <summary>Mask coefficients, for a segmentation head. Empty for a plain detector.</summary>
    public double[] Seg = [];

    public Box Clone() => new()
    {
        X1 = X1, Y1 = Y1, X2 = X2, Y2 = Y2, Conf = Conf, Cls = Cls, Label = Label,
        Seg = (double[])Seg.Clone(),
    };
}

public sealed record DetectResult(List<Box> Boxes) : IResult;

/// <summary>Which suppression the model's output type asks for.</summary>
public enum NmsMode
{
    ClassAgnostic,
    PerClass,
}

/// <summary>
/// Decodes an anchors-first YOLO output and suppresses overlaps.
///
/// <para>
/// One type with an <see cref="NmsMode"/> field rather than a base class and a subclass overriding
/// the suppression — CONVENTIONS §5. The reference has <c>PerClassYOLODetectorPostprocessing</c>
/// inheriting from the plain one; flattening it removes a place where Go's non-virtual embedding
/// would silently call the wrong method, and reads the same in all four languages.
/// </para>
/// </summary>
public sealed class YoloDetector(string[] labels, double iou, double cls, NmsMode mode)
    : IPostprocessor
{
    /// <summary>
    /// When set, the decode stops before truncation and labelling.
    ///
    /// <para>
    /// The segmentation path needs the RAW float coordinates and the mask coefficients — truncating
    /// first loses sub-pixel information the mask crop depends on.
    /// </para>
    /// </summary>
    public bool NumpyOnly { get; private init; }

    /// <summary>
    /// A copy with raw-coordinate mode on.
    ///
    /// <para>
    /// Exists so the loader switch stays the ONE place a detector is constructed: the segmentation
    /// model needs the same detector with one flag flipped, and building a second one there would
    /// duplicate the argument list that the switch already owns.
    /// </para>
    /// </summary>
    public YoloDetector WithNumpyOnly() =>
        new(labels, iou, cls, mode) { NumpyOnly = true };

    public IResult Apply(NdArray output, Context context) =>
        new DetectResult(Decode(output, context));

    public List<Box> Decode(NdArray output, Context context)
    {
        ReadOnlySpan<float> data = output.AsFloat32();
        int[] shape = output.Shape;
        if (shape.Length == 3 && shape[0] == 1)
        {
            shape = shape[1..];
        }
        if (shape.Length != 2)
        {
            throw new InvalidDataException(
                $"postprocess: detector expects [anchors, 4+nc(+seg)], got " +
                $"{NdArray.Describe(output.Shape)}");
        }

        int anchors = shape[0], stride = shape[1];
        int nc = labels.Length;
        if (stride < 4 + nc)
        {
            throw new InvalidDataException(
                $"postprocess: row width {stride} is too small for {nc} classes");
        }
        int segLen = stride - 4 - nc;

        var boxes = new List<Box>(64);
        for (int a = 0; a < anchors; a++)
        {
            ReadOnlySpan<float> row = data.Slice(a * stride, stride);

            // Strict `>` for the best class, matching np.argmax's first-maximum rule.
            int best = 0;
            double bestScore = double.NegativeInfinity;
            for (int c = 0; c < nc; c++)
            {
                if (row[4 + c] > bestScore)
                {
                    best = c;
                    bestScore = row[4 + c];
                }
            }
            // `!(score > cls)` rather than `score <= cls`: identical for real numbers, and it keeps
            // the reference's own spelling, which also handles NaN the same way.
            if (!(bestScore > cls))
            {
                continue;
            }

            double cx = row[0], cy = row[1], w = row[2], h = row[3];
            var box = new Box
            {
                X1 = cx - w / 2, Y1 = cy - h / 2,
                X2 = cx + w / 2, Y2 = cy + h / 2,
                Conf = bestScore, Cls = best,
            };
            if (segLen > 0)
            {
                box.Seg = new double[segLen];
                for (int i = 0; i < segLen; i++)
                {
                    box.Seg[i] = row[4 + nc + i];
                }
            }
            boxes.Add(box);
        }

        if (boxes.Count == 0)
        {
            return [];
        }

        List<int> keep = mode == NmsMode.PerClass
            ? NmsPerClass(boxes, iou)
            : Nms([.. Enumerable.Range(0, boxes.Count)], boxes, iou);

        var kept = keep.Select(i => boxes[i]).ToList();

        // **Stable sort, reading order.** LINQ's OrderBy is stable; List.Sort is NOT. Two boxes with
        // the same y1 must keep the order suppression left them in, because that order decides which
        // word comes first in a joined field string.
        kept = [.. kept.OrderBy(b => b.Y1).ThenBy(b => b.X1)];

        if (context.Resize)
        {
            MapBack(kept, context);
        }

        if (!NumpyOnly)
        {
            foreach (Box b in kept)
            {
                // TRUNCATION, not rounding — the reference casts to int here, after having rounded
                // half-to-even in MapBack. Two different operations, in that order.
                b.X1 = Math.Truncate(b.X1);
                b.Y1 = Math.Truncate(b.Y1);
                b.X2 = Math.Truncate(b.X2);
                b.Y2 = Math.Truncate(b.Y2);
                b.Label = b.Cls >= 0 && b.Cls < labels.Length
                    ? labels[b.Cls]
                    : "Unsupported class";
            }
        }
        return kept;
    }

    /// <summary>
    /// Undoes the letterbox and the extra padding, then clamps and rounds.
    ///
    /// <para>
    /// The order matters and is the reference's: unpad, unscale, un-extra-pad, clamp negatives to
    /// zero, THEN round half-to-even. Rounding before clamping would turn -0.4 into -0 and then 0,
    /// which happens to agree; rounding before unscaling would not.
    /// </para>
    /// </summary>
    private static void MapBack(List<Box> boxes, Context ctx)
    {
        foreach (Box b in boxes)
        {
            b.X1 = (b.X1 - ctx.PadLetter[0]) / ctx.Ratio - ctx.PadExtra[0];
            b.X2 = (b.X2 - ctx.PadLetter[0]) / ctx.Ratio - ctx.PadExtra[0];
            b.Y1 = (b.Y1 - ctx.PadLetter[1]) / ctx.Ratio - ctx.PadExtra[1];
            b.Y2 = (b.Y2 - ctx.PadLetter[1]) / ctx.Ratio - ctx.PadExtra[1];

            b.X1 = Math.Max(0, b.X1);
            b.Y1 = Math.Max(0, b.Y1);
            b.X2 = Math.Max(0, b.X2);
            b.Y2 = Math.Max(0, b.Y2);
            b.Conf = Math.Max(0, b.Conf);
            for (int j = 0; j < b.Seg.Length; j++)
            {
                // **Clamps the WHOLE row, mask coefficients included.** It reads like an oversight —
                // the intent was surely to clamp coordinates — but the reference does
                // `detect_res[detect_res < 0] = 0` across every column, and those coefficients go
                // straight to the segmentor. A port that clamps only the coordinates gets different
                // masks.
                b.Seg[j] = Math.Max(0, b.Seg[j]);
            }

            b.X1 = PyNum.RoundHalfEven(b.X1);
            b.Y1 = PyNum.RoundHalfEven(b.Y1);
            b.X2 = PyNum.RoundHalfEven(b.X2);
            b.Y2 = PyNum.RoundHalfEven(b.Y2);
            b.Conf = Ops.RoundHalfEven(b.Conf, 3);
        }
    }

    /// <summary>
    /// Non-maximum suppression over a subset of indices.
    ///
    /// <para>
    /// **The tie rule is load-bearing.** The candidates are sorted ASCENDING by score with a stable
    /// sort, and the highest — the LAST element — is picked each round. On a score tie that keeps the
    /// one with the greater original index, which is what <c>np.argsort</c> plus <c>order[-1]</c>
    /// does. Picking the first instead changes which of two equally-confident boxes survives.
    /// </para>
    ///
    /// <para><c>ratio &lt; threshold</c> survives, matching <c>np.where(ratio &lt; threshold)</c>.</para>
    /// </summary>
    private static List<int> Nms(List<int> indices, List<Box> boxes, double threshold)
    {
        var areas = indices.ToDictionary(i => i,
            i => (boxes[i].X2 - boxes[i].X1) * (boxes[i].Y2 - boxes[i].Y1));

        // Stable ascending by score — np.argsort.
        var order = indices.OrderBy(i => boxes[i].Conf).ToList();

        var picked = new List<int>();
        while (order.Count > 0)
        {
            int index = order[^1];
            picked.Add(index);
            order.RemoveAt(order.Count - 1);

            var survivors = new List<int>(order.Count);
            foreach (int j in order)
            {
                double x1 = Math.Max(boxes[index].X1, boxes[j].X1);
                double y1 = Math.Max(boxes[index].Y1, boxes[j].Y1);
                double x2 = Math.Min(boxes[index].X2, boxes[j].X2);
                double y2 = Math.Min(boxes[index].Y2, boxes[j].Y2);
                double inter = Math.Max(0, x2 - x1) * Math.Max(0, y2 - y1);
                double ratio = inter / (areas[index] + areas[j] - inter);
                if (ratio < threshold)
                {
                    survivors.Add(j);
                }
            }
            order = survivors;
        }
        return picked;
    }

    /// <summary>
    /// Suppression within each class independently.
    ///
    /// <para>
    /// Classes are visited in ASCENDING order, matching <c>np.unique</c>. The resulting index order
    /// feeds the reading-order sort afterwards, so a different visit order can break ties
    /// differently — which is why this sorts the class list rather than iterating a set.
    /// </para>
    /// </summary>
    private static List<int> NmsPerClass(List<Box> boxes, double threshold)
    {
        var classes = boxes.Select(b => b.Cls).Distinct().OrderBy(c => c).ToList();
        var keep = new List<int>();
        foreach (int c in classes)
        {
            var indices = Enumerable.Range(0, boxes.Count).Where(i => boxes[i].Cls == c).ToList();
            keep.AddRange(Nms(indices, boxes, threshold));
        }
        return keep;
    }
}

File: ports/java/docproc/src/main/kotlin/net/russiandocs/docproc/postprocess/Detector.kt
package net.russiandocs.docproc.postprocess

import net.russiandocs.docproc.tensors.NdArray
import net.russiandocs.docproc.tensors.Ops
import net.russiandocs.docproc.tensors.PyNum
import kotlin.math.max
import kotlin.math.min
import kotlin.math.truncate

/** One detection. Mutable, because the coordinates are rewritten in place when mapped back. */
public class Box {
    public var x1: Double = 0.0
    public var y1: Double = 0.0
    public var x2: Double = 0.0
    public var y2: Double = 0.0
    public var conf: Double = 0.0
    public var cls: Int = 0
    public var label: String = ""

    /** Mask coefficients, for a segmentation head. Empty for a plain detector. */
    public var seg: DoubleArray = DoubleArray(0)

    public fun copy(): Box = Box().also {
        it.x1 = x1; it.y1 = y1; it.x2 = x2; it.y2 = y2
        it.conf = conf; it.cls = cls; it.label = label
        it.seg = seg.copyOf()
    }
}

public data class DetectResult(val boxes: List<Box>) : ModelResult

/** Which suppression the model's output type asks for. */
public enum class NmsMode {
    CLASS_AGNOSTIC,
    PER_CLASS,
}

/**
 * Decodes an anchors-first YOLO output and suppresses overlaps.
 *
 * One type with an [NmsMode] field rather than a base class and a subclass overriding the suppression —
 * CONVENTIONS §5. The reference has `PerClassYOLODetectorPostprocessing` inheriting from the plain one;
 * flattening it removes a place where Go's non-virtual embedding would silently call the wrong method, and
 * reads the same in all four languages.
 */
public class YoloDetector(
    private val labels: List<String>,
    private val iou: Double,
    private val cls: Double,
    private val mode: NmsMode,
    /**
     * When set, the decode stops before truncation and labelling.
     *
     * The segmentation path needs the RAW float coordinates and the mask coefficients — truncating first
     * loses sub-pixel information the mask crop depends on.
     */
    public val numpyOnly: Boolean = false,
) : Postprocessor {

    /**
     * A copy with raw-coordinate mode on.
     *
     * Exists so the loader switch stays the ONE place a detector is constructed: the segmentation model
     * needs the same detector with one flag flipped, and building a second one there would duplicate the
     * argument list the switch already owns.
     */
    public fun withNumpyOnly(): YoloDetector = YoloDetector(labels, iou, cls, mode, numpyOnly = true)

    override fun apply(output: NdArray, context: Context): ModelResult =
        DetectResult(decode(output, context))

    public fun decode(output: NdArray, context: Context): List<Box> {
        val data = output.asFloat32()
        var shape = output.shape
        if (shape.size == 3 && shape[0] == 1) {
            shape = shape.copyOfRange(1, shape.size)
        }
        require(shape.size == 2) {
            "postprocess: detector expects [anchors, 4+nc(+seg)], got " +
                NdArray.describe(output.shape)
        }

        val anchors = shape[0]
        val stride = shape[1]
        val nc = labels.size
        require(stride >= 4 + nc) {
            "postprocess: row width $stride is too small for $nc classes"
        }
        val segLen = stride - 4 - nc

        val boxes = ArrayList<Box>(64)
        for (a in 0 until anchors) {
            val base = a * stride

            // Strict `>` for the best class, matching np.argmax's first-maximum rule.
            var best = 0
            var bestScore = Double.NEGATIVE_INFINITY
            for (c in 0 until nc) {
                val score = data[base + 4 + c].toDouble()
                if (score > bestScore) {
                    best = c
                    bestScore = score
                }
            }
            // `!(score > cls)` rather than `score <= cls`: identical for real numbers, and it keeps the
            // reference's own spelling, which also handles NaN the same way.
            if (!(bestScore > cls)) {
                continue
            }

            val cx = data[base].toDouble()
            val cy = data[base + 1].toDouble()
            val w = data[base + 2].toDouble()
            val h = data[base + 3].toDouble()
            val box = Box().also {
                it.x1 = cx - w / 2
                it.y1 = cy - h / 2
                it.x2 = cx + w / 2
                it.y2 = cy + h / 2
                it.conf = bestScore
                it.cls = best
            }
            if (segLen > 0) {
                box.seg = DoubleArray(segLen) { i -> data[base + 4 + nc + i].toDouble() }
            }
            boxes += box
        }

        if (boxes.isEmpty()) {
            return emptyList()
        }

        val keep = if (mode == NmsMode.PER_CLASS) {
            nmsPerClass(boxes, iou)
        } else {
            nms(boxes.indices.toList(), boxes, iou)
        }

        // **Stable sort, reading order.** `sortedWith` is stable on the JVM; `Collections.sort` on a
        // mutable list is too, but an unstable primitive sort is not. Two boxes with the same y1 must keep
        // the order suppression left them in, because that order decides which word comes first in a
        // joined field string.
        var kept = keep.map { boxes[it] }
        kept = kept.sortedWith(compareBy({ it.y1 }, { it.x1 }))

        if (context.resize) {
            mapBack(kept, context)
        }

        if (!numpyOnly) {
            for (b in kept) {
                // TRUNCATION, not rounding — the reference casts to int here, after having rounded
                // half-to-even in mapBack. Two different operations, in that order.
                b.x1 = truncate(b.x1)
                b.y1 = truncate(b.y1)
                b.x2 = truncate(b.x2)
                b.y2 = truncate(b.y2)
                b.label = if (b.cls in labels.indices) labels[b.cls] else "Unsupported class"
            }
        }
        return kept
    }

    /**
     * Undoes the letterbox and the extra padding, then clamps and rounds.
     *
     * The order matters and is the reference's: unpad, unscale, un-extra-pad, clamp negatives to zero,
     * THEN round half-to-even.
     */
    private fun mapBack(boxes: List<Box>, ctx: Context) {
        for (b in boxes) {
            b.x1 = (b.x1 - ctx.padLetter[0]) / ctx.ratio - ctx.padExtra[0]
            b.x2 = (b.x2 - ctx.padLetter[0]) / ctx.ratio - ctx.padExtra[0]
            b.y1 = (b.y1 - ctx.padLetter[1]) / ctx.ratio - ctx.padExtra[1]
            b.y2 = (b.y2 - ctx.padLetter[1]) / ctx.ratio - ctx.padExtra[1]

            b.x1 = max(0.0, b.x1)
            b.y1 = max(0.0, b.y1)
            b.x2 = max(0.0, b.x2)
            b.y2 = max(0.0, b.y2)
            b.conf = max(0.0, b.conf)
            for (j in b.seg.indices) {
                // **Clamps the WHOLE row, mask coefficients included.** It reads like an oversight — the
                // intent was surely to clamp coordinates — but the reference does
                // `detect_res[detect_res < 0] = 0` across every column, and those coefficients go straight
                // to the segmentor. A port that clamps only the coordinates gets different masks.
                b.seg[j] = max(0.0, b.seg[j])
            }

            b.x1 = PyNum.roundHalfEven(b.x1)
            b.y1 = PyNum.roundHalfEven(b.y1)
            b.x2 = PyNum.roundHalfEven(b.x2)
            b.y2 = PyNum.roundHalfEven(b.y2)
            b.conf = Ops.roundHalfEven(b.conf, 3)
        }
    }

    /**
     * Non-maximum suppression over a subset of indices.
     *
     * **The tie rule is load-bearing.** The candidates are sorted ASCENDING by score with a stable sort,
     * and the highest — the LAST element — is picked each round. On a score tie that keeps the one with the
     * greater original index, which is what `np.argsort` plus `order[-1]` does. Picking the first instead
     * changes which of two equally-confident boxes survives.
     *
     * `ratio < threshold` survives, matching `np.where(ratio < threshold)`.
     */
    private fun nms(indices: List<Int>, boxes: List<Box>, threshold: Double): List<Int> {
        val areas = indices.associateWith { i ->
            (boxes[i].x2 - boxes[i].x1) * (boxes[i].y2 - boxes[i].y1)
        }

        // Stable ascending by score — np.argsort.
        var order = indices.sortedBy { boxes[it].conf }

        val picked = ArrayList<Int>()
        while (order.isNotEmpty()) {
            val index = order.last()
            picked += index
            order = order.dropLast(1)

            order = order.filter { j ->
                val x1 = max(boxes[index].x1, boxes[j].x1)
                val y1 = max(boxes[index].y1, boxes[j].y1)
                val x2 = min(boxes[index].x2, boxes[j].x2)
                val y2 = min(boxes[index].y2, boxes[j].y2)
                val inter = max(0.0, x2 - x1) * max(0.0, y2 - y1)
                val ratio = inter / (areas.getValue(index) + areas.getValue(j) - inter)
                ratio < threshold
            }
        }
        return picked
    }

    /**
     * Suppression within each class independently.
     *
     * Classes are visited in ASCENDING order, matching `np.unique`. The resulting index order feeds the
     * reading-order sort afterwards, so a different visit order can break ties differently — which is why
     * this sorts the class list rather than iterating a set.
     */
    private fun nmsPerClass(boxes: List<Box>, threshold: Double): List<Int> {
        val classes = boxes.map { it.cls }.distinct().sorted()
        val keep = ArrayList<Int>()
        for (c in classes) {
            keep += nms(boxes.indices.filter { boxes[it].cls == c }, boxes, threshold)
        }
        return keep
    }
}
